R/utilities-object-functions.R

```diff
@@ -0,0 +1,248 @@
+    }
+    ierr <- which(colnames(obs) == err) # only one
+    if (length(ierr) == 0) {
+      stop(paste("Column with error estimates not found in observations", err))
+    }
+  }
+
+  ## The dependent variables
+  type <- 1 # data input type: type 2 is table format, type 1 is long format...
+
+  if (!is.null(y)) { # it is in table format; first column are names of observed data...
+
+    Names <- as.character(unique(obs[, 1])) # Names of data sets, all data should be model variables...
+    Ndat <- length(Names) # Number of data sets
+    ilist <- 1:Ndat
+    if (!is.character(y)) {
+      stop("'y' should be the *name* of the column with the values of the dependent variable in obs")
+    }
+    iy <- which(colnames(obs) == y)
+    if (length(iy) == 0) {
+      stop(paste("Column with value of dependent variable not found in observations", y))
+    }
+    type <- 2
+  } else { # it is a matrix, variable names are column names
+    Ndat <- NCOL(obs) - 1
+    Names <- colnames(obs)
+    ilist <- (1:NCOL(obs)) # column positions of the (dependent) observed variables
+    exclude <- ix # exclude columns that are not
+    if (ierr > 0) {
+      exclude <- c(ix, ierr)
+    } # exclude columns that are not
+    if (length(exclude) > 0) {
+      ilist <- ilist[-exclude]
+    }
+  }
+
+  # ================================
+  # The model results
+  # ================================
+
+  ModNames <- colnames(model) # Names of model variables
+  if (length(ix) > 1) {
+    ixMod <- NULL
+
+    for (i in 1:length(ix)) {
+      ix2 <- which(colnames(model) == x[i])
+      if (length(ix2) == 0) {
+        stop(paste("Cannot calculate cost: independent variable not found in model output", x[i]))
+      }
+      ixMod <- c(ixMod, ix2)
+    }
+
+    xMod <- model[, ixMod] # Independent variable, model
+  } else if (length(ix) == 1) {
+    ixMod <- which(colnames(model) == x)
+    if (length(ixMod) == 0) {
+      stop(paste("Cannot calculate cost: independent variable not found in model output", x))
+    }
+    xMod <- model[, ixMod] # Independent variable, model
+  }
+  Residual <- NULL
+  CostVar <- NULL
+
+  # ================================
+  # Compare model and data...
+  # ================================
+  xDat <- 0
+  iDat <- 1:nrow(obs)
+
+  for (i in ilist) { # for each observed variable ...
+    ii <- which(ModNames == Names[i])
+    if (length(ii) == 0) stop(paste("observed variable not found in model output", Names[i]))
+    yMod <- model[, ii]
+    if (type == 2) { # table format
+      iDat <- which(obs[, 1] == Names[i])
+      if (length(ix) > 0) xDat <- obs[iDat, ix]
+      obsdat <- obs[iDat, iy]
+    } else {
+      if (length(ix) > 0) xDat <- obs[, 1]
+      obsdat <- obs[, i]
+    }
+    ii <- which(is.na(obsdat))
+    if (length(ii) > 0) {
+      xDat <- xDat[-ii]
+      obsdat <- obsdat[-ii]
+    }
+
+    if (length(ix) > 0) {
+      # Only interpolate if more than one x point is available
+      if (length(unique(xMod)) > 1) {
```

Comment:
This was before:

`ModVar <- approx(xMod, yMod, xout = xDat)$y`

Now only interpolate if there are more than one simulated x point. Otherwise sort the simulated values based on the order of observed x values